x-pack/plugins/alerting/server/saved_objects/migrations.ts

```diff
@@ -902,3 +930,56 @@ function pipeMigrations(...migrations: AlertMigration[]): AlertMigration {
   return (doc: SavedObjectUnsanitizedDoc<RawRule>) =>
     migrations.reduce((migratedDoc, nextMigration) => nextMigration(migratedDoc), doc);
 }
+
+function mapSearchSourceMigrationFunc(
+  migrateSerializedSearchSourceFields: MigrateFunction<SerializedSearchSourceFields>
+): MigrateFunction {
+  return (state) => {
+    const _state = state as { attributes: RawRule };
+
+    const serializedSearchSource = _state.attributes.params.searchConfiguration;
+
+    if (isSerializedSearchSource(serializedSearchSource)) {
+      return {
+        ..._state,
+        attributes: {
+          ..._state.attributes,
+          params: {
+            ..._state.attributes.params,
+            searchConfiguration: migrateSerializedSearchSourceFields(serializedSearchSource),
+          },
+        },
+      };
+    }
+    return _state;
+  };
+}
+
+/**
+ * This creates a migration map that applies search source migrations to legacy es query rules.
+ * It doesn't modify existing migrations. The following migrations will occur at minimum version of 8.3+.
+ */
+function getSearchSourceMigrations(
+  encryptedSavedObjects: EncryptedSavedObjectsPluginSetup,
+  searchSourceMigrations: MigrateFunctionsObject
+) {
+  const filteredMigrations: SavedObjectMigrationMap = {};
+  for (const versionKey in searchSourceMigrations) {
+    if (versionKey >= MINIMUM_SS_MIGRATION_VERSION) {
+      const migrateSearchSource = mapSearchSourceMigrationFunc(
+        searchSourceMigrations[versionKey]
+      ) as unknown as AlertMigration;
+
+      filteredMigrations[versionKey] = executeMigrationWithErrorHandling(
+        createEsoMigration(
+          encryptedSavedObjects,
+          (doc: SavedObjectUnsanitizedDoc<RawRule>): doc is SavedObjectUnsanitizedDoc<RawRule> =>
+            true,
```

Comment:
Synced with @gmmorris, dear @Dmitriynj  here is where we should check for `isEsQueryRuleType` to prevent unnecessary decryption 
FYI: @mikecote & @pmuellr

Reply:
Done.